### PR TITLE
[BREAKING] Introduce block.chainid

### LINKED ISSUE #8854: Expose chainid in the language
The `chainid` opcode was introduced in Istanbul and added as an opcode to inline assembly a while ago (see #7310).

It would be worth considering adding language support to it. I can imagine three options:
1) `chainid()` like `gasleft()`
2) `tx.chainid`
3) `block.chainid`

While I think 2) is nicer because less pollution of the global namespace, the [EIP](https://eips.ethereum.org/EIPS/eip-1344) specifically says this property is part of the "chain configuration" of the client and not the transaction. It was discussed to make per-transaction, but that was decided against.

Option 3) may be the best considering the above, while the chainid property not being part of the block, it can only ever change on a block basis (in case of forking a chain).

Motivation: chanid is used via [ERC-712](https://eips.ethereum.org/EIPS/eip-712) and [ERC-1612](https://github.com/ethereum/EIPs/issues/2613).

More motivation: reduce the need for using assembly for such a feature, this also helps in improved the likelyhood of SMTChecker succeeding in more contracts. (One example is [Uniswap](https://github.com/Uniswap/uniswap-v2-core/blob/master/contracts/UniswapV2ERC20.sol#L27))

### PR DESCRIPTION
**Pull request by axic**
Closes #8854.
Depends on #10558.

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -2922,6 +2922,14 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
 		)
 			annotation.isPure = true;
+		else if (magicType->kind() == MagicType::Kind::Block && memberName == "chainid" && !m_evmVersion.hasChainID())
+		{
+			m_errorReporter.typeError(
+				3081_error,
+				_memberAccess.location(),
+				"\"chainid\" is not supported by the VM version."
+			);
+		}
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
			m_errorReporter.typeError(
				3081_error,
				_memberAccess.location(),
				"\"chainid\" is not supported by the VM version."
			);
```

File: libsolidity/analysis/ViewPureChecker.cpp [hunk 2]
```diff
@@ -368,7 +368,6 @@ void ViewPureChecker::endVisit(MemberAccess const& _memberAccess)
 			{MagicType::Kind::ABI, "encodePacked"},
 			{MagicType::Kind::ABI, "encodeWithSelector"},
 			{MagicType::Kind::ABI, "encodeWithSignature"},
-			{MagicType::Kind::Block, "blockhash"},
```
Comment (chriseth) [inline libsolidity/analysis/ViewPureChecker.cpp:371]: Should be mentioned in the changelog.

**Comment by axic** [comment]
@leonardoalt Test failures are SMTChecker counterexample issues...

**Comment by leonardoalt** [comment]
Yea I'm looking into it

**Comment by leonardoalt** [comment]
Did this PR change anything in the AST?

**Comment by axic** [comment]
> Did this PR change anything in the AST?

Should not.

File: libsolidity/analysis/ViewPureChecker.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/analysis/ViewPureChecker.cpp:371]: This is dead code. Can remove from this PR and deal with this sperately.

Comment (chriseth) [reply libsolidity/analysis/ViewPureChecker.cpp:371]: Ah sorry!

**Review by chriseth — APPROVED** [review]

**Pull request merged**